### ISSUE TITLE
Certificate (/etc/ssl/certs/java/cacerts) issue in Java images on OpenShift

### ISSUE THREAD

**Issue by chanseokoh** [GoogleContainerTools/distroless#456]
A user on the distroless Slack channel reported that Java images have a certificate issue. This is probably due to #372.

**Comment by loosebazooka** [comment]
revert for now then? and do a release?

**Comment by chanseokoh** [comment]
I thought it was a permission issue, but all the directories and the file is world-readable. I'll revert this now and look into it.

**Comment by loosebazooka** [comment]
Cool. Nice work catching this.

**Comment by chanseokoh** [comment]
I cannot reproduce it, and everything looks normal. However, I've reverted #372 with #457 for precaution. The new images will go live soon automatically.

**Comment by chanseokoh** [comment]
Customer account:

> I assumed it worked with root, because it works when running tests with Docker only, but fails in Openshift.
> so I added a test trying to list all the files in/etc/ssl/certs/java/ on application startup, and got an "java.nio.file.AccessDeniedException: /etc/ssl/certs/java"
> this also just started happening recently (monday I think), which it why I assumed those changes were the cause
> will test it again with the latest image

> can confirm that everything is working again with the latest java 11 image

@chrismcbride any idea? I couldn't find any problem. Permissions looked OK, and the `cacerts` looked OK.

**Comment by chanseokoh** [comment]
I believe OpenShift runs a container as a random UID, if this helps.

**Comment by chrismcbride** [comment]
Huh...I have no clue. I built a java8 debian debug image with my change:

```
/ # ls -alh /etc/ssl/certs/java/cacerts 
-rw-r--r--    1 1000     100       177.9K Jan  1  1970 /etc/ssl/certs/java/cacerts
```

And then without my change...

```
/ # ls -alh /etc/ssl/certs/java/cacerts 
-rw-r--r--    1 1000     100       173.1K Jan  1  1970 /etc/ssl/certs/java/cacerts
```
Looks identical besides a slight difference in file size...which may just be from the alias names I use for the certs but that shouldnt have any affect. Im stumped...I dont know much about openshift though.

**Comment by chrismcbride** [comment]
ah nvm I see the problem. The "java" directory is owned by root and is in the group root in my change, but its UID 1000, group 100 before my change. I can work on a fix later.

**Comment by chanseokoh** [comment]
I don't get it. When I pull and check the latest images (with the PR reverted), every directory and `cacerts` are owned by `root:root`.

```
docker run --rm -it --entrypoint /busybox/sh gcr.io/distroless/java:11-debug
Unable to find image 'gcr.io/distroless/java:11-debug' locally
11-debug: Pulling from distroless/java
9ff2acc3204b: Already exists 
69e2f037cdb3: Already exists 
1a3f070d750b: Already exists 
3e010093287c: Already exists 
367901af3297: Pull complete 
Digest: sha256:e91e23383a8843a3f0bb00bdf99b9b7bed8c7ce25ed929acbe0eedec70fa91f9
Status: Downloaded newer image for gcr.io/distroless/java:11-debug
/ # ls -ld etc/
drwxr-xr-x    1 root     root          4096 Jan 23 17:11 etc/
/ # ls -ld etc/ssl/
drwxr-xr-x    1 root     root          4096 Jan  1  1970 etc/ssl/
/ # ls -ld etc/ssl/certs/
drwxr-xr-x    1 root     root          4096 Jan  1  1970 etc/ssl/certs/
/ # ls -ld etc/ssl/certs/java/
drwxr-xr-x    2 root     root          4096 Jan  1  1970 etc/ssl/certs/java/
/ # ls -ld etc/ssl/certs/java/cacerts 
-rw-r--r--    1 root     root        177280 Jan  1  1970 etc/ssl/certs/java/cacerts
/ # 
```

**Comment by chrismcbride** [comment]
I ran the following commands:
```
git checkout 76c76a13d6bed32db25b88c2f2ff2ea52df52c82
bazel build //java:java8_debug_debian9.tar
docker load -i bazel-bin/java/java8_debug_debian9.tar
docker run --rm -it --entrypoint=/busybox/sh bazel-bin/java/java8_debug_debian9.tar
ls -al /etc/ssl/certs/
```

Ran the same commands again except with 01579c3dd0f9b4397fe4f50769a9caea130b15e7 (pre-revert). On the first I got:

```
/ # ls -al /etc/ssl/certs/
total 244
drwxr-xr-x    1 1000     100           4096 Jan  1  1970 .
drwxr-xr-x    1 1000     100           4096 Jan  1  1970 ..
-rw-r--r--    1 1000     100         235192 Jan  1  1970 ca-certificates.crt
drwxr-xr-x    2 1000     100           4096 Jan  1  1970 java
```

On the second I got:
```
/ # ls -al /etc/ssl/certs/
total 244
drwxr-xr-x    1 root     root          4096 Jan 23 16:14 .
drwxr-xr-x    1 root     root          4096 Jan 23 16:14 ..
-rw-r--r--    1 1000     100         235192 Jan  1  1970 ca-certificates.crt
drwxr-xr-x    2 root     root          4096 Jan 23 16:14 java
```

**Comment by chanseokoh** [comment]
@chrismcbride I made an interesting observation. The UID and GID of these directories are those of the host OS. For example, my UID:GID is `252384:89939`. It's the same when you do `bazel run //java:java8_debug_debian9` to push directly to Docker. Surprising! I actually didn't know.

```
-rw-r--r--    1 252384   89939       235192 Jan  1  1970 ca-certificates.crt                           
drwxr-xr-x    2 252384   89939         4096 Jan  1  1970 java                                                                                                                                                 
```

In any cases, the official images on GCR have `root:root`.

Still no clue as to why the user has a problem on OpenShift.

**Comment by chrismcbride** [comment]
ah yea thats makes sense. I also dont know why its failing for openshift then...

**Comment by chrismcbride** [comment]
@chanseokoh Do you have the exact steps to reproduce? Is it possible the openshift user wasnt using the published images on gcr but was building them from bazel?

**Comment by chanseokoh** [comment]
@chrismcbride the user has provided this [test program](https://github.com/andersflemmen/distroless-test). Details are in the README of the repo. The user also reported that running as non-root locally with Docker works. So there may be something special with OpenShift. The user also said if they manually supply and overwrite the cacerts file (by placing it at `src/main/jib/etc/ssl/certs/java/cacerts` which then Jib will pick up and put in the image), it works again.

I tried running my own test application on CodeReady Container (which they say "brings a minimal, preconfigured OpenShift 4.1 or newer cluster to your local laptop or desktop computer for development and testing"), but it seems to run the application as root. I think I'll try [OpenShift Online] later.

**Comment by chanseokoh** [comment]
> Is it possible the openshift user wasnt using the published images on gcr but was building them from bazel?

In case you missed it, the test program builds the image using the reverted image from GCR: https://github.com/andersflemmen/distroless-test/blob/master/pom.xml#L27

**Comment by chanseokoh** [comment]
So I tried OpenShift Online as I said, and unlike CodeReady Containers, I can see it run a container as non-root; `System.getProperty("user.name")` returns some large number. However, I couldn't reproduce the issue. I think this is unrelated to root vs. non-root. Actually there has been no evidence that it is due to non-root. The user reported that locally, Docker run works whether it is root or not. And it works on my OpenShift Online instance where the container runs as non-root. There may be something special about the user's OpenShift instance.

```
System.getProperty("user.name")=1015300000
Verifying certificates...
/etc/ssl/certs/java exists, listing files:/etc/ssl/certs/java/cacerts
Certificates verified OK!
                                                                                                                                                                                          
                                                                                                                                                                                          
Getting https://google.com
Got response from https://google.com with status 200 OK
```

The base I used in this case was `gcr.io/distroless/java@sha256:cc00de1dfd7141dd7d3be0cf1da8e1147f912e0f088550c0ddc0cbe14a70c27c` (https://github.com/GoogleContainerTools/distroless/issues/456#issuecomment-577779398)

**Comment by chanseokoh** [comment]
For deploying to OpenShift Online, I first did the traditional `kubectl apply -f ....yaml`, but I also tested the same command as the user used: `oc run my-test --image=francium25/test3` (Interestingly, the `user.name` is always 1015300000, but I think this isn't important.)

**Comment by chanseokoh** [comment]
I think I figured it out.

When examining the layer with the `cacerts` file in the current image (by `docker save ...`), the layer tarball includes parent directories for the file.

```
$ tar tvf layer.tar | grep etc/ssl
tar: Removing leading `/' from member names
drwxr-xr-x 0/0               0 1969-12-31 19:00 ./etc/ssl/
drwxr-xr-x 0/0               0 1969-12-31 19:00 ./etc/ssl/certs/
drwxr-xr-x 0/0               0 1969-12-31 19:00 ./etc/ssl/certs/java/
-rw-r--r-- 0/0          177280 1969-12-31 19:00 ./etc/ssl/certs/java/cacerts
lrwxrwxrwx 0/0               0 1969-12-31 19:00 ./usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/cacerts -> /etc/ssl/certs/java/cacerts
```

However, the layer tarball built with the reverted change (#372) has an entry for the file only and no entries for parent directories.

```
$ tar tvf layer.tar | grep etc/ssl
tar: Removing leading `/' from member names
-rw-r--r-- 0/0          182217 1969-12-31 19:00 ./etc/ssl/certs/java/cacerts
lrwxrwxrwx 0/0               0 1969-12-31 19:00 ./usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/cacerts -> /etc/ssl/certs/java/cacerts
```
As such, `dive` doesn't (and can't) show the permissions of some of the parent directories for which the image had never had an entry.
```
drwxr-xr-x        0:0     182 kB  │   │   └── ssl
drwxr-xr-x        0:0     182 kB  │   │       └── certs
----------        0:0     182 kB  │   │           └── java
-rw-r--r--        0:0     182 kB  │   │               └── cacerts
```
In the end, I think it will be up to a runtime to determine the permissions of parents directories that have no entries in layer tars. This also explains why some runtimes (like Docker and my OpenShift online cluster) have no issue. I'll put up a fix shortly.

**Comment by chanseokoh** [comment]
Will be fixed by #478. The user confirmed on the Distroless Slack channel that #478 fixes the original issue.

**Comment by chrismcbride** [comment]
oh nice find! thanks for digging in to this
